<a href="https://colab.research.google.com/github/khariharan897/BronzeBadge_GENAI_Assessment/blob/main/Assignment1_EmployeeProductivity_LR.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Assignment 1:
A company wants to predict employee productivity scores to improve workforce planning and training programs. You are hired as a Data Scientist to build a multivariate linear regression model that predicts an employee’s Productivity Score based on multiple work-related factors.

Experience (yrs),Training Hours,Working Hours,Projects,Productivity Score

2,40,38,3,62

5,60,42,6,78

1,20,35,2,55

8,80,45,8,88

4,50,40,5,72

10,90,48,9,92

3,30,37,4,65

6,70,44,7,82

7,75,46,7,85

2,25,36,3,60

Interpretation

Which factor most strongly impacts productivity?

How does training affect productivity?

Should the company increase training hours or working hours?

What happens if Working Hours increase beyond optimal limits?

Can productivity ever decrease with more experience?

How would you detect overfitting in this model?

Suggest one new feature to improve prediction accuracy.

In [ ]:

# Employee Productivity Prediction using Multivariate Linear Regression
# Import pandas to create and work with DataFrames

import pandas as pd
import io


# Create the dataset using a Python dictionary.

data = {
    "Experience": [2, 5, 1, 8, 4, 10, 3, 6, 7, 2],
    "TrainingHours": [40, 60, 20, 80, 50, 90, 30, 70, 75, 25],
    "WorkingHours": [38, 42, 35, 45, 40, 48, 37, 44, 46, 36],
    "Projects": [3, 6, 2, 8, 5, 9, 4, 7, 7, 3],
    "Productivity": [62, 78, 55, 88, 72, 92, 65, 82, 85, 60]
}

# Convert the dictionary into a DataFrame
df = pd.DataFrame(data)

display(df.head())

,Experience,TrainingHours,WorkingHours,Projects,Productivity
0,2,40,38,3,62
1,5,60,42,6,78
2,1,20,35,2,55
3,8,80,45,8,88
4,4,50,40,5,72


In [ ]:
# Display the complete dataset
print("Employee Dataset")
print(df)

Employee Dataset
   Experience  TrainingHours  WorkingHours  Projects  Productivity
0           2             40            38         3            62
1           5             60            42         6            78
2           1             20            35         2            55
3           8             80            45         8            88
4           4             50            40         5            72
5          10             90            48         9            92
6           3             30            37         4            65
7           6             70            44         7            82
8           7             75            46         7            85
9           2             25            36         3            60


In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

# Define features (X) and target (y)
X = df[['Experience', 'TrainingHours', 'WorkingHours', 'Projects']]
y = df['Productivity']

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize and train the Linear Regression model
model = LinearRegression()
model.fit(X_train, y_train)

print("Model training complete. Coefficients and intercept are available.")

Model training complete. Coefficients and intercept are available.


In [ ]:
from sklearn.metrics import r2_score

# Get the coefficients and feature names
coefficients = pd.DataFrame({'Feature': X.columns, 'Coefficient': model.coef_})
display(coefficients)

# Get the intercept
print(f"Intercept: {model.intercept_:.2f}")

# Make predictions on the test set
y_pred = model.predict(X_test)

# Evaluate the model
r2 = r2_score(y_test, y_pred)
print(f"R-squared on test set: {r2:.2f}")

,Feature,Coefficient
0,Experience,-0.263658
1,TrainingHours,0.272684
2,WorkingHours,-1.085511
3,Projects,4.885986


Intercept: 78.17
R-squared on test set: 0.31


In [ ]:

# Predict productivity scores
predicted = model.predict(X_test)

print("\nActual Productivity")
print(list(y_test))

print("\nPredicted Productivity")
print(predicted)


Actual Productivity
[85, 78]

Predicted Productivity
[81.04275534 76.93586698]


In [ ]:

# Import evaluation metrics
from sklearn.metrics import mean_squared_error, r2_score
# Calculate Mean Squared Error
mse = mean_squared_error(y_test, predicted)

# Calculate R2 Score
r2 = r2_score(y_test, predicted)

print("\nModel Evaluation")

print("Mean Squared Error =", round(mse, 2))
print("R2 Score =", round(r2, 2))


Model Evaluation
Mean Squared Error = 8.4
R2 Score = 0.31


In [ ]:

# Create details of a new employee
new_employee = pd.DataFrame(
    [[7, 55, 40, 7]],
    columns=["Experience", "TrainingHours", "WorkingHours", "Projects"]
)

# Predict productivity
new_prediction = model.predict(new_employee)

print("\nPredicted Productivity for New Employee")
print(round(new_prediction[0], 2))


Predicted Productivity for New Employee
82.1


Interpretation questions:

1) Which factor most strongly impacts productivity?

    Based on the absolute magnitude of the coefficients, Projects (4.89) has the strongest positive impact on productivity. For each additional project an employee undertakes, their productivity score is predicted to increase by approximately 4.89 units, holding other factors constant.

2) How does training affect productivity?
    The coefficient for Training Hours is 0.27. This indicates a positive relationship: for every additional hour of training, the productivity score is predicted to increase by 0.27 units, assuming other factors remain constant.

3) Should the company increase training hours or working hours?

    Given the model results, the company should increase training hours. An increase in 'Training Hours' is associated with an increase in productivity (coefficient 0.27). Conversely, an increase in 'Working Hours' is associated with a decrease in productivity (coefficient -1.09).

4) What happens if Working Hours increase beyond optimal limits?
    
    The current linear model suggests that increasing 'Working Hours' continuously decreases productivity (coefficient -1.09). However, a simple linear model may not fully capture real-world complexities. In reality, there might be an optimal number of working hours, after which productivity sharply declines due to factors like burnout or fatigue. A linear model assumes this negative relationship holds indefinitely, which is likely unrealistic beyond certain limits. More advanced models or polynomial terms might be needed to capture such non-linear effects.

5) Can productivity ever decrease with more experience?

    According to this model, yes, productivity is predicted to decrease with more experience, as the coefficient for 'Experience (yrs)' is -0.26. This is counter-intuitive in many contexts and might suggest that other unmeasured factors (confounding variables) are at play, or that there's a non-linear relationship that this model isn't capturing. It could also be due to the small sample size or specific characteristics of this dataset.

6) How would you detect overfitting in this model?
 Overfitting occurs when a model learns the training data too well, performing poorly on new, unseen data. To detect overfitting, one would typically:

    Compare training and test performance:     A significantly higher R-squared on the training set compared to the test set (our test R-squared is 0.31) would indicate overfitting.
    Cross-validation: Use techniques like K-fold cross-validation to get a more robust estimate of the model's performance on various subsets of the data, providing a better average performance metric and indicating stability.
    Analyze residuals: Look for patterns in the residuals (differences between actual and predicted values) that might suggest the model is fitting noise rather than the underlying pattern.

7) Suggest one new feature to improve prediction accuracy.

    A new feature that could potentially improve prediction accuracy is Job Satisfaction. Higher job satisfaction often correlates with increased motivation, engagement, and ultimately, higher productivity. Incorporating this as a quantitative measure (e.g., from employee surveys) could help explain more variance in the productivity scores.